docs/proposals/templates.md

```diff
@@ -0,0 +1,350 @@
+be available within the cluster for a plugin feels overly burdensome.  A map between the `templateKind` and
+the api endpoint capable of expanding it is necessary, however.
+
+Template processors could also be provided in client form to help template authors perform local testing+debugging.
+
+Nested templates should be handled by the primary endpoint, retrieving and handing each nested template off to the appropriate
+expansion plugin, and then assembling the aggregate list of objects from the top level template and any nested templates.
+
+## Open discussion points
+### Where to define parameters (closed?)
+
+There has been some discussion around where to define parameters that are being injected into a Template
+
+1. In a separate standalone file
+2. Within the Template itself
+
+This proposal suggests including the parameter definitions within the Template, which provides a self-contained structure that
+can be easily versioned, transported, and instantiated without risk of mismatching content.
+
+Separating the parameter definitions makes for a more complex process with respect to
+* Editing a template (if/when first class editing tools are created)
+* Storing/retrieving template objects with a central store
+
+### Where to define parameters (closed?)
+
+This question should now be closed as the problem is delegated to individual template expansion plugins.  They will
+be provided with a map of parameter values and metadata and can inject/substitute those parameters into the 
+template body based on the scheme they define.
+
+### Where to store templates
+
+It has been suggested that k8s should enable template storage with the benefits being:
+
+* Allows the user to see and share available templates
+* Provides unified access control
+  * Storing templates elsewhere means enabling k8s to access that store (eg github) which means providing credentials
+    and more potential points of configuration error for users.
+* Simplified service catalog flows in which instantiating a service is equivalent to instantiating a template
+* Templates can still be stored elsewhere for versioning/revision control, with the currently accepted version
+  residing in the k8s store for use.
+
+However there are also concerns that storing templates doesn't fit with the primary mission of the k8s etcd store,
+which is to store the definitions of objects being actively managed/run by k8s.
+
+### Supporting generated values
+
+There is a use case for being able to define a parameter value that should have an auto-generated value on
+instantiation (eg passwords).  How do we want to support this?  It would be preferable to have this be part of
+the parameter definition if we can agree on a universally suitable approach.  Generation can still be delegated
+to the template expansion plugin (it would receive all metadata about the parameter, including generator inputs),
```

Comment:
@smarterclayton and I discussed moving generation to Secret using annotations and initializers (#3585), since that's the main use case, which I like better than having it in Template.